dm/dm/master/scheduler/scheduler.go

```diff
@@ -693,12 +720,51 @@ func (s *Scheduler) TransferSource(source, worker string) error {
 	if err2 := s.updateStatusToBound(w, ha.NewSourceBound(source, worker)); err2 != nil {
 		s.logger.DPanic("we have checked w.stage is free, so there should not be an error", zap.Error(err2))
 	}
-
-	// 6. try bound the old worker
+	// 6. now this old worker is free, try bound source to it
 	_, err = s.tryBoundForWorker(oldWorker)
 	if err != nil {
 		s.logger.Warn("in transfer source, error when try bound the old worker", zap.Error(err))
 	}
+	s.mu.Unlock()
+	return nil
+}
+
+// batchOperateTaskOnWorker batch operate tasks in one worker and use query-status to make sure all tasks are in expected stage if needWait=true.
+func (s *Scheduler) batchOperateTaskOnWorker(
+	ctx context.Context, worker *Worker, tasks []string, stage pb.Stage, needWait bool) error {
+	// pause this running tasks
+	for _, taskName := range tasks {
+		if err := s.UpdateExpectSubTaskStage(stage, taskName, worker.bound.Source); err != nil {
+			return err
+		}
+	}
+	if !needWait {
+		return nil
+	}
+	// wait all tasks are in expected stage before actually starting scheduling
+WaitLoop:
+	for maxRetryNum := 10; maxRetryNum > 0; maxRetryNum-- {
+		resp, err := worker.queryStatus(ctx)
+		if err != nil {
+			return terror.Annotatef(err, "failed to query worker: %s status", worker.baseInfo.Name)
+		}
+		for _, status := range resp.QueryStatus.GetSubTaskStatus() {
+			if status.Stage != pb.Stage_Paused {
+				// NOTE: the defaultRPCTimeout is 10m, use 1s * retry times to increase the waiting time
+				sleepTime := time.Second * time.Duration(10-maxRetryNum)
+				s.logger.Info(
+					"waiting task",
+					zap.String("task", status.Name),
+					zap.Int("retry times", 10-maxRetryNum),
+					zap.Duration("sleep time", sleepTime),
+					zap.String("want stage", stage.String()),
+					zap.String("current stage", status.Stage.String()),
+				)
+				time.Sleep(sleepTime)
+				goto WaitLoop
```

Comment:
this retry loop is totally wrong...

when any task is paused, break the inner loop. when no task is paused, break outer loop

Reply:
emm, why? i think what i need is : `all tast in paused stage` 🤔, so when one task not in `paused` i wait and break the inner loop and goto outer loop

Reply:
https://go.dev/play/p/sEFHr1O5nOw

Reply:
sorry i misuse `goto`..... fixed and add more test in  https://github.com/pingcap/tiflow/commit/86a7a4d3d1e1527701c06517c992e721fa3ba71d